flang/lib/Lower/PFTBuilder.cpp

```diff
@@ -1027,6 +1045,31 @@ struct SymbolDependenceDepth {
       std::vector<std::vector<lower::pft::Variable>> &vars)
       : vars{vars} {}
 
+  // Analyze the equivalence sets. This analysis need not be performed when the
+  // scope has no equivalence sets.
+  void analyzeAliases(const semantics::Scope &scope) {
+    IntervalSet::Allocator allocator;
+    IntervalSet intervals(allocator);
+
+    // Collect the offset ranges which have aliasing.
+    for (const auto &set : scope.equivalenceSets())
+      for (const auto &eqv : set) {
+        const auto &sym = eqv.symbol;
+        aliasSyms.insert(&sym);
+        intervals.merge(sym.offset(), sym.offset() + sym.size() - 1);
+      }
+
```

Comment:
The `IntervalSet` is pretty nice ! It took me a while to understand why you were not simply using `start = max(sym.offset, start); stop = max(sym.offset + sym.size, stop);` for each equivalence set and creating one store per equivalence set. I think you could add a note that different equivalence set can overlap if that is indeed the rational for using `IntervalSet`. That is not obvious for non Fortran programmers  (at least it was not obvious for me).

Reply:
It is to deal with partially overlapping address ranges, which is a feature of EQUIVALENCE.  The algorithm also allows LLVM code gen to generate the best layout and code for the target rather than imposing a rigid top-down stack frame structure.

Reply:
For example, we can have `integer :: a(10), b(10), c(10)` with `equivalence (a(9),b(1)), (b(9),c(1))` and create one storage sequence of 26 integers and 8 nominal locations that participate in unions.